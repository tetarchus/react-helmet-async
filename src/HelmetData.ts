import { mapStateOnServer } from './server';

import type { HelmetDataContext, HelmetServerState } from './types';

const domInstances = [];

const clearInstances = (): void => {
  domInstances.length = 0;
};

class HelmetData {
  public readonly instances = [];

  public canUseDOM: boolean;

  public context: HelmetDataContext;

  public readonly value = {
    setHelmet: (serverState: HelmetServerState): void => {
      this.context.helmet = serverState;
    },
    helmetInstances: {
      get: () => (this.canUseDOM ? domInstances : this.instances),
      add: (instance): void => {
        (this.canUseDOM ? domInstances : this.instances).push(instance);
      },
      remove: (instance): void => {
        const index = (this.canUseDOM ? domInstances : this.instances).indexOf(instance);
        (this.canUseDOM ? domInstances : this.instances).splice(index, 1);
      },
    },
  };

  public constructor(context: HelmetDataContext, canUseDOM = typeof document !== 'undefined') {
    this.context = context;
    this.canUseDOM = canUseDOM;

    if (!canUseDOM) {
      this.context.helmet = mapStateOnServer({
        baseTag: [],
        bodyAttributes: {},
        encodeSpecialCharacters: true,
        htmlAttributes: {},
        linkTags: [],
        metaTags: [],
        noscriptTags: [],
        scriptTags: [],
        styleTags: [],
        title: '',
        titleAttributes: {},
      });
    }
  }
}

export { clearInstances, HelmetData };
