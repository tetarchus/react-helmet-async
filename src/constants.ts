const TAG_PROPERTIES = {
  CHARSET: 'charset',
  CSS_TEXT: 'cssText',
  HREF: 'href',
  HTTPEQUIV: 'http-equiv',
  INNER_HTML: 'innerHTML',
  ITEM_PROP: 'itemprop',
  NAME: 'name',
  PROPERTY: 'property',
  REL: 'rel',
  SRC: 'src',
};

const ATTRIBUTE_NAMES = {
  BODY: 'bodyAttributes',
  HTML: 'htmlAttributes',
  TITLE: 'titleAttributes',
} as const;

const TAG_NAMES = {
  BASE: 'base',
  BODY: 'body',
  FRAGMENT: 'Symbol(react.fragment)',
  HEAD: 'head',
  HTML: 'html',
  LINK: 'link',
  META: 'meta',
  NOSCRIPT: 'noscript',
  SCRIPT: 'script',
  STYLE: 'style',
  TITLE: 'title',
} as const;

const SELF_CLOSING_TAGS = [TAG_NAMES.NOSCRIPT, TAG_NAMES.SCRIPT, TAG_NAMES.STYLE] as const;

const SEO_PRIORITY_TAGS = {
  link: { rel: ['amphtml', 'canonical', 'alternate'] },
  script: { type: ['application/ld+json'] },
  meta: {
    charset: '',
    name: ['robots', 'description'],
    property: [
      'og:type',
      'og:title',
      'og:url',
      'og:image',
      'og:image:alt',
      'og:description',
      'twitter:url',
      'twitter:title',
      'twitter:description',
      'twitter:image',
      'twitter:image:alt',
      'twitter:card',
      'twitter:site',
    ],
  },
};

const VALID_TAG_NAMES = Object.keys(TAG_NAMES).map((tagName) => TAG_NAMES[tagName] as string);

const REACT_TAG_MAP = {
  accesskey: 'accessKey',
  charset: 'charSet',
  class: 'className',
  contenteditable: 'contentEditable',
  contextmenu: 'contextMenu',
  'http-equiv': 'httpEquiv',
  itemprop: 'itemProp',
  tabindex: 'tabIndex',
};

const HTML_TAG_MAP = Object.fromEntries(
  Object.keys(REACT_TAG_MAP).map((key) => [REACT_TAG_MAP[key as keyof typeof REACT_TAG_MAP], key]),
);

const HELMET_ATTRIBUTE = 'data-rh' as const;

const HELMET_PROPS = {
  DEFAULT_TITLE: 'defaultTitle',
  DEFER: 'defer',
  ENCODE_SPECIAL_CHARACTERS: 'encodeSpecialCharacters',
  ON_CHANGE_CLIENT_STATE: 'onChangeClientState',
  TITLE_TEMPLATE: 'titleTemplate',
  PRIORITIZE_SEO_TAGS: 'prioritizeSeoTags',
} as const;

export {
  ATTRIBUTE_NAMES,
  HELMET_ATTRIBUTE,
  HELMET_PROPS,
  HTML_TAG_MAP,
  REACT_TAG_MAP,
  SELF_CLOSING_TAGS,
  SEO_PRIORITY_TAGS,
  TAG_NAMES,
  TAG_PROPERTIES,
  VALID_TAG_NAMES,
};
