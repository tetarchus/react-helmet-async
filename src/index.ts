export { Helmet } from './Helmet';
export { HelmetData } from './HelmetData';
export { HelmetContextProvider } from './HelmetContext';

export type { HelmetContext, HelmetProps, HelmetProviderProps } from './types';
