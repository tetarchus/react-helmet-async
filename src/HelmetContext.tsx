import React from 'react';

import { HelmetData } from './HelmetData';

import type { HelmetContextValue, HelmetProviderProps } from './types';

const defaultValue: HelmetContextValue = {
  canUseDOM: typeof document !== 'undefined',
  context: {},
};

const HelmetContext: React.Context<HelmetContextValue> = React.createContext(defaultValue);

const HelmetContextProvider: React.FC<HelmetProviderProps> = ({
  children,
  context = defaultValue.context,
  ssr,
}: HelmetProviderProps) => {
  const { canUseDOM: contextCUD } = React.useContext(HelmetContext);
  // Allows override to 'canUseDOM'
  const canUseDOM = ssr != null ? !ssr : contextCUD;
  const contextValue = React.useMemo(
    () => ({ canUseDOM, context: new HelmetData(context, canUseDOM).value }),
    [canUseDOM, context],
  );

  return <HelmetContext.Provider value={contextValue}>{children}</HelmetContext.Provider>;
};

export { HelmetContext, HelmetContextProvider };
