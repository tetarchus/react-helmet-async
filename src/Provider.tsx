import React from 'react';

import { HelmetData } from './HelmetData';

const defaultValue = {};

const Context = React.createContext(defaultValue);

const canUseDOM = typeof document !== 'undefined';
class Provider extends React.Component {
  static canUseDOM = canUseDOM;

  static defaultProps = {
    context: {},
  };

  static displayName = 'HelmetProvider';

  constructor(props) {
    super(props);

    this.helmetData = new HelmetData(this.props.context, Provider.canUseDOM);
  }

  render() {
    return <Context.Provider value={this.helmetData.value}>{this.props.children}</Context.Provider>;
  }
}

export { Context, Provider };
