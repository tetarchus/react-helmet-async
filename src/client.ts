import { getDocument } from 'ssr-window';

import { HELMET_ATTRIBUTE, TAG_NAMES, TAG_PROPERTIES } from './constants';
import { flattenArray, isArray } from './utils';

import type {
  ArrayType,
  CheckedUpdatedTags,
  HelmetProps,
  HelmetState,
  ObjectValues,
  UpdatedTags,
} from './types';

const doc = getDocument();

const checkRemoval = (
  oldTags: Element[],
  newTags: Element[],
  newElement:
    | HTMLBaseElement
    | HTMLElement
    | HTMLLinkElement
    | HTMLMetaElement
    | HTMLScriptElement
    | HTMLStyleElement,
): CheckedUpdatedTags => {
  let indexToDelete = -1;
  const updatedOldTags = [...oldTags];
  const updatedNewTags = [...newTags];
  const match = updatedOldTags.some((existingTag, index) => {
    indexToDelete = index;
    return newElement.isEqualNode(existingTag);
  });
  if (match && indexToDelete > -1) {
    updatedOldTags.splice(indexToDelete, 1);
  } else {
    updatedNewTags.push(newElement);
  }

  return { updatedOldTags, updatedNewTags };
};

const updateTags = <
  T extends ObjectValues<
    Pick<typeof TAG_NAMES, 'BASE' | 'LINK' | 'META' | 'NOSCRIPT' | 'SCRIPT' | 'STYLE'>
  >,
>(
  type: T,
  tags: HelmetProps[T],
): UpdatedTags => {
  const headElement = doc.head;
  const tagNodes = headElement.querySelectorAll(`${type}[${HELMET_ATTRIBUTE}]`);
  let oldTags = Array.prototype.slice.call(tagNodes) as Element[];
  let newTags = [] as Element[];

  if (tags != null && 'length' in tags && isArray(tags) && tags.length > 0) {
    for (const tag of tags) {
      const currentTag = tag as ArrayType<NonNullable<HelmetProps[T]>>;
      const newElement = document.createElement(type);

      for (const attribute of Object.keys(currentTag)) {
        if (Object.prototype.hasOwnProperty.call(currentTag, attribute)) {
          if (attribute === TAG_PROPERTIES.INNER_HTML) {
            newElement.innerHTML = currentTag.innerHTML ?? '';
          } else if (attribute === TAG_PROPERTIES.CSS_TEXT) {
            if (newElement.styleSheet != null) {
              newElement.styleSheet.cssText = (currentTag.cssText as string | undefined) ?? '';
            } else {
              newElement.append(document.createTextNode(currentTag.cssText));
            }
          } else {
            const value = (currentTag[attribute] as string | undefined) ?? '';
            newElement.setAttribute(attribute, value);
          }
        }
      }

      newElement.setAttribute(HELMET_ATTRIBUTE, 'true');

      // Remove a duplicate tag from domTagstoRemove, so it isn't cleared.
      const { updatedOldTags, updatedNewTags } = checkRemoval(oldTags, newTags, newElement);
      oldTags = [...updatedOldTags];
      newTags = [...updatedNewTags];
    }
  }

  for (const tag of oldTags) tag.remove();
  for (const tag of newTags) headElement.append(tag);

  return {
    oldTags,
    newTags,
  };
};

const updateAttributes = <T extends ObjectValues<typeof TAG_NAMES>>(
  tagName: T,
  attributes: HelmetState['bodyAttributes'] | HelmetState['htmlAttributes'],
): void => {
  const elementTag = document.querySelectorAll(tagName)[0] as Element | undefined;
  const elementAttributes = { ...attributes };

  if (elementTag != null) {
    const helmetAttributeString = elementTag.getAttribute(HELMET_ATTRIBUTE);
    const helmetAttributes = helmetAttributeString != null ? helmetAttributeString.split(',') : [];
    const attributesToRemove = [...helmetAttributes];
    const attributeKeys = Object.keys(elementAttributes);

    for (const attribute of attributeKeys) {
      const value = (elementAttributes[attribute] as string | undefined) ?? '';

      if (elementTag.getAttribute(attribute) !== value) {
        elementTag.setAttribute(attribute, value);
      }

      if (!helmetAttributes.includes(attribute)) {
        helmetAttributes.push(attribute);
      }

      const indexToSave = attributesToRemove.indexOf(attribute);
      if (indexToSave !== -1) {
        attributesToRemove.splice(indexToSave, 1);
      }
    }

    for (let i = attributesToRemove.length - 1; i >= 0; i -= 1) {
      elementTag.removeAttribute(attributesToRemove[i]);
    }

    if (helmetAttributes.length === attributesToRemove.length) {
      elementTag.removeAttribute(HELMET_ATTRIBUTE);
    } else if (elementTag.getAttribute(HELMET_ATTRIBUTE) !== attributeKeys.join(',')) {
      elementTag.setAttribute(HELMET_ATTRIBUTE, attributeKeys.join(','));
    }
  }
};

const updateTitle = (
  title: HelmetState['title'],
  attributes: HelmetState['titleAttributes'],
): void => {
  if (typeof title !== 'undefined' && document.title !== title) {
    document.title = flattenArray(title);
  }

  updateAttributes(TAG_NAMES.TITLE, attributes);
};

const commitTagChanges = (newState: HelmetState, cb?: () => void): void => {
  const {
    baseTag,
    bodyAttributes,
    htmlAttributes,
    linkTags,
    metaTags,
    noscriptTags,
    onChangeClientState,
    scriptTags,
    styleTags,
    title,
    titleAttributes,
  } = newState;
  updateAttributes(TAG_NAMES.BODY, bodyAttributes);
  updateAttributes(TAG_NAMES.HTML, htmlAttributes);

  updateTitle(title, titleAttributes);

  const tagUpdates = {
    baseTag: updateTags(TAG_NAMES.BASE, baseTag),
    linkTags: updateTags(TAG_NAMES.LINK, linkTags),
    metaTags: updateTags(TAG_NAMES.META, metaTags),
    noscriptTags: updateTags(TAG_NAMES.NOSCRIPT, noscriptTags),
    scriptTags: updateTags(TAG_NAMES.SCRIPT, scriptTags),
    styleTags: updateTags(TAG_NAMES.STYLE, styleTags),
  };

  const addedTags = {};
  const removedTags = {};

  for (const tagType of Object.keys(tagUpdates)) {
    const { newTags, oldTags } = tagUpdates[tagType] as UpdatedTags;

    if (newTags.length > 0) {
      addedTags[tagType] = newTags;
    }
    if (oldTags.length > 0) {
      removedTags[tagType] = oldTags;
    }
  }

  cb?.();

  onChangeClientState?.(newState, addedTags, removedTags);
};

let _helmetCallback: number | null = null;

const handleStateChangeOnClient = (newState: HelmetState): void => {
  if (_helmetCallback != null) {
    cancelAnimationFrame(_helmetCallback);
  }

  if (newState.defer ?? false) {
    _helmetCallback = requestAnimationFrame(() => {
      commitTagChanges(newState, () => {
        _helmetCallback = null;
      });
    });
  } else {
    commitTagChanges(newState);
    _helmetCallback = null;
  }
};

export { handleStateChangeOnClient };
