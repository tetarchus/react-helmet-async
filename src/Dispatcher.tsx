import React from 'react';

import { handleStateChangeOnClient } from './client';
import { Provider } from './Provider';
import { mapStateOnServer } from './server';
import { reducePropsToState } from './utils';

import type { DispatcherProps, HelmetServerState } from './types';

const Dispatcher: React.FC<DispatcherProps> = ({ context }: DispatcherProps) => {
  const [rendered, setRendered] = React.useState(false);

  // componentWillUnmount() {
  //   const { helmetInstances } = this.props.context;
  //   helmetInstances.remove(this);
  //   this.emitChange();
  // }

  const emitChange = React.useCallback(() => {
    const { helmetInstances, setHelmet } = context;
    let serverState: HelmetServerState | null = null;
    const state = reducePropsToState(helmetInstances.get().map((instance) => ({ ...instance })));
    if (Provider.canUseDOM) {
      handleStateChangeOnClient(state);
    } else if (mapStateOnServer) {
      serverState = mapStateOnServer(state);
    }
    setHelmet(serverState);

    return () => {
      const { helmetInstances } = context;
      helmetInstances.remove(this);
    };
  }, []);

  const init = React.useCallback(() => {
    if (rendered) {
      return;
    }

    setRendered(true);

    const { helmetInstances } = context;
    helmetInstances.add(this);
    emitChange();
  }, []);

  init();

  return null;
};

export { Dispatcher };
