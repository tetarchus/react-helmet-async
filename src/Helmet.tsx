import React from 'react';

import { Dispatcher } from './Dispatcher';
import { HelmetContext } from './HelmetContext';
import { HelmetData } from './HelmetData';
import { mapChildrenToProps } from './utils';

import type { HelmetProps, HelmetPropsWithoutChildren } from './types';

const Helmet: React.FC<HelmetProps> = ({
  base,
  bodyAttributes,
  children,
  defaultTitle,
  defer = true,
  encodeSpecialCharacters = true,
  helmetData,
  htmlAttributes,
  link,
  meta,
  noscript,
  onChangeClientState,
  prioritizeSeoTags = false,
  script,
  style,
  title,
  titleAttributes,
  titleTemplate,
}: HelmetProps) => {
  let newProps: HelmetPropsWithoutChildren = {
    base,
    bodyAttributes,
    defaultTitle,
    defer,
    encodeSpecialCharacters,
    helmetData,
    htmlAttributes,
    link,
    meta,
    noscript,
    onChangeClientState,
    prioritizeSeoTags,
    script,
    style,
    title,
    titleAttributes,
    titleTemplate,
  };
  let data = helmetData;

  if (children != null) {
    newProps = mapChildrenToProps(children, newProps);
  }

  if (data != null && !(data instanceof HelmetData)) {
    data = new HelmetData(data.context);
  }

  return data instanceof HelmetData ? (
    <Dispatcher {...newProps} context={data.value} helmetData={undefined} />
  ) : (
    <HelmetContext.Consumer>
      {(value): JSX.Element => <Dispatcher {...newProps} context={value.context} />}
    </HelmetContext.Consumer>
  );
};

export { Helmet };
