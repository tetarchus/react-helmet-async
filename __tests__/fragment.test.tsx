import React from 'react';

import { Helmet } from '../src';

import { render } from './setup/testSetup';

describe('React.Fragments', () => {
  it('parses Fragments', () => {
    expect.assertions(1);
    render(
      <Helmet defer={false}>
        <>
          <title>Hello</title>
          <meta charSet='utf-8' />
        </>
      </Helmet>,
    );

    expect(document.title).toMatchSnapshot();
  });

  it('parses nested Fragments', () => {
    expect.assertions(1);

    render(
      <Helmet defer={false}>
        <>
          <title>Foo</title>
          <>
            <title>Bar</title>
            <title>Baz</title>
          </>
        </>
      </Helmet>,
    );

    expect(document.title).toMatchSnapshot();
  });
});
