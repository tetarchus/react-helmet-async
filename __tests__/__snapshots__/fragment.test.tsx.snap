// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React.Fragments parses Fragments 1`] = `"Hello"`;

exports[`React.Fragments parses nested Fragments 1`] = `"Baz"`;
